pkg/ca/intermediateca/intermediateca.go

```diff
@@ -37,6 +42,88 @@ type IntermediateCA struct {
 	Signer crypto.Signer
 }
 
+func (ica *IntermediateCA) CreatePrecertificate(ctx context.Context, challenge *challenges.ChallengeResult) (*ca.CodeSigningPreCertificate, error) {
+	cert, err := x509ca.MakeX509(challenge)
+	if err != nil {
+		return nil, err
+	}
+
+	certChain, privateKey := ica.getX509KeyPair()
+
+	// Append poison extension
+	cert.ExtraExtensions = append(cert.ExtraExtensions, pkix.Extension{
+		Id:       asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 11129, 2, 4, 3},
+		Critical: true,
+		Value:    asn1.NullBytes,
+	})
+
+	finalCertBytes, err := x509.CreateCertificate(rand.Reader, cert, certChain[0], challenge.PublicKey, privateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	csc, err := ca.CreateCSCFromDER(challenge, finalCertBytes, certChain)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ca.CodeSigningPreCertificate{
+		Subject:    csc.Subject,
+		PreCert:    csc.FinalCertificate,
+		CertChain:  csc.FinalChain,
+		PrivateKey: privateKey,
+	}, nil
+}
+
+// From https://github.com/letsencrypt/boulder/blob/54b697d51b9f63cfd6055577cd317d4096aeab08/issuance/issuance.go#L497
+func generateSCTListExt(scts []ct.SignedCertificateTimestamp) (pkix.Extension, error) {
+	list := ctx509.SignedCertificateTimestampList{}
+	for _, sct := range scts {
+		sctBytes, err := cttls.Marshal(sct)
+		if err != nil {
+			return pkix.Extension{}, err
+		}
+		list.SCTList = append(list.SCTList, ctx509.SerializedSCT{Val: sctBytes})
+	}
+	listBytes, err := cttls.Marshal(list)
+	if err != nil {
+		return pkix.Extension{}, err
+	}
+	extBytes, err := asn1.Marshal(listBytes)
+	if err != nil {
+		return pkix.Extension{}, err
+	}
+	return pkix.Extension{
+		Id:    asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 11129, 2, 4, 2},
+		Value: extBytes,
+	}, nil
+}
+
+func (ica *IntermediateCA) IssueFinalCertificate(ctx context.Context, precert *ca.CodeSigningPreCertificate, sct *ct.SignedCertificateTimestamp) (*ca.CodeSigningCertificate, error) {
+	// remove poison extension from precertificate.
```

Comment:
did you look at using https://github.com/google/certificate-transparency-go/blob/58784bdeda2ec153f22c23ff1d6cfcc6f1c7f5a2/x509/x509.go#L1208 with `precert.PreCert.RawTBSCertificate`? there is a fair bit of logic in the ct-library but your code probably covers most of the same stuff.